identity-store/persistence/cqldb/persist.go

```diff
@@ -0,0 +1,227 @@
+	if p.err != nil {
+		return &iterator{err: p.err}
+	}
+
+	var b strings.Builder
+	b.WriteString(cqldb.SelectCommand + " ")
+	b.WriteString(ownerKey)
+	b.WriteString(",")
+	b.WriteString(deviceIDKey)
+	b.WriteString(" " + cqldb.FromClause + " ")
+	b.WriteString(p.table)
+	b.WriteString(" " + cqldb.WhereClause + " ")
+	b.WriteString(ownerKey)
+	b.WriteString("='")
+	b.WriteString(owner)
+	b.WriteString("'")
+
+	iter := p.tx.Query(b.String()).WithContext(p.ctx).Iter()
+
+	return &iterator{
+		iter: iter,
+		ctx:  p.ctx,
+	}
+}
+
+// RetrieveAll retrieves all user's authorized devices.
+func (p *PersistenceTx) RetrieveAll() persistence.Iterator {
+	if p.err != nil {
+		return &iterator{err: p.err}
+	}
+
+	var b strings.Builder
+	b.WriteString(cqldb.SelectCommand + " ")
+	b.WriteString(ownerKey)
+	b.WriteString(",")
+	b.WriteString(deviceIDKey)
+	b.WriteString(" " + cqldb.FromClause + " ")
+	b.WriteString(p.table)
+
+	iter := p.tx.Query(b.String()).WithContext(p.ctx).Iter()
+
+	return &iterator{
+		iter: iter,
+		ctx:  p.ctx,
+	}
+}
+
+type iterator struct {
+	err  error
+	iter *gocql.Iter
+	ctx  context.Context
+}
+
+func (i *iterator) Next(s *persistence.AuthorizedDevice) bool {
+	if i.err != nil {
+		return false
+	}
+	return i.iter.Scan(&s.Owner, &s.DeviceID)
+}
+
+func (i *iterator) Err() error {
+	return i.err
+}
+
+func (i *iterator) Close() {
+	if i.iter != nil {
+		i.err = i.iter.Close()
+	}
+}
+
+// Persist device's authorization details.
+func (p *PersistenceTx) Persist(d *persistence.AuthorizedDevice) error {
+	if p.err != nil {
+		return p.err
+	}
+
+	data, ok, err := p.RetrieveByDevice(d.DeviceID)
+	if err != nil {
+		return err
+	}
+	if ok {
+		if data.Owner != d.Owner {
+			return fmt.Errorf("device %v already exists", d.DeviceID)
+		}
+		return nil
+	}
+
+	var b strings.Builder
+	b.WriteString("INSERT INTO ")
+	b.WriteString(p.table)
+	b.WriteString(" (")
+	b.WriteString(ownerKey)
+	b.WriteString(",")
+	b.WriteString(deviceIDKey)
+	b.WriteString(") VALUES (")
+	b.WriteString("'")
+	b.WriteString(d.Owner)
+	b.WriteString("'")
+	b.WriteString(",")
+	b.WriteString(d.DeviceID)
+	b.WriteString(") IF NOT EXISTS;")
+
+	return p.tx.Query(b.String()).WithContext(p.ctx).Exec()
+}
+
+// Delete removes the device's authorization record.
+func (p *PersistenceTx) Delete(deviceID, userID string) error {
+	if p.err != nil {
+		return p.err
+	}
+
+	var b strings.Builder
+	b.WriteString("delete from ")
+	b.WriteString(p.table)
+	b.WriteString(" " + cqldb.WhereClause + " ")
+	b.WriteString(ownerKey)
+	b.WriteString("=")
+	b.WriteString("'")
+	b.WriteString(userID)
+	b.WriteString("'")
+	b.WriteString(" and ")
+	b.WriteString(deviceIDKey)
+	b.WriteString("=")
+	b.WriteString(deviceID)
+
+	return p.tx.Query(b.String()).WithContext(p.ctx).Exec()
+}
+
+// Close must always be called (use defer immediately after NewTransaction).
+func (p *PersistenceTx) Close() {
+	// do nothing - transaction is not supported
```

Comment:
So these are not transactions? Can't that be an issue?

Reply:
Fixed